experimental/kernels/gpt2_webgpu_aot.cpp

```diff
@@ -0,0 +1,1088 @@
+    std::vector<Tensor> fcb; // (L, 4*C)
+    std::vector<Tensor> fcprojw; // (L, C, 4*C)
+    std::vector<Tensor> fcprojb; // (L, C)
+    Tensor lnfw; // (C)
+    Tensor lnfb; // (C)
+} ParameterTensors;
+
+void fill_in_parameter_sizes(size_t* param_sizes, GPT2Config config) {
+    size_t Vp = config.padded_vocab_size;
+    size_t C = config.channels;
+    size_t maxT = config.max_seq_len;
+    size_t L = config.num_layers;
+    param_sizes[0] = Vp * C; // wte
+    param_sizes[1] = maxT * C; // wpe
+    param_sizes[2] = L * C; // ln1w
+    param_sizes[3] = L * C; // ln1b
+    param_sizes[4] = L * (3 * C) * C; // qkvw
+    param_sizes[5] = L * (3 * C); // qkvb
+    param_sizes[6] = L * C * C; // attprojw
+    param_sizes[7] = L * C; // attprojb
+    param_sizes[8] = L * C; // ln2w
+    param_sizes[9] = L * C; // ln2b
+    param_sizes[10] = L * (4 * C) * C; // fcw
+    param_sizes[11] = L * (4 * C); // fcb
+    param_sizes[12] = L * C * (4 * C); // fcprojw
+    param_sizes[13] = L * C; // fcprojb
+    param_sizes[14] = C; // lnfw
+    param_sizes[15] = C; // lnfb
+}
+
+// allocate memory for the parameters and point the individual tensors to the right places
+void malloc_and_point_parameters(Context& ctx, ParameterTensors* params, size_t* param_sizes) {
+    params->wte = createTensor(ctx, Shape{param_sizes[0]}, kf32);
+    params->wpe = createTensor(ctx, Shape{param_sizes[1]}, kf32);
+    for(int l = 0; l < NUM_PARAMETER_LAYERS; l++) {
+      params->ln1w.push_back(createTensor(ctx, Shape{param_sizes[2]/NUM_PARAMETER_LAYERS}, kf32));
+      params->ln1b.push_back(createTensor(ctx, Shape{param_sizes[3]/NUM_PARAMETER_LAYERS}, kf32));
+      params->qkvw.push_back(createTensor(ctx, Shape{param_sizes[4]/NUM_PARAMETER_LAYERS}, kf32));
+      params->qkvb.push_back(createTensor(ctx, Shape{param_sizes[5]/NUM_PARAMETER_LAYERS}, kf32));
+      params->attprojw.push_back(createTensor(ctx, Shape{param_sizes[6]/NUM_PARAMETER_LAYERS}, kf32));
+      params->attprojb.push_back(createTensor(ctx, Shape{param_sizes[7]/NUM_PARAMETER_LAYERS}, kf32));
+      params->ln2w.push_back(createTensor(ctx, Shape{param_sizes[8]/NUM_PARAMETER_LAYERS}, kf32));
+      params->ln2b.push_back(createTensor(ctx, Shape{param_sizes[9]/NUM_PARAMETER_LAYERS}, kf32));
+      params->fcw.push_back(createTensor(ctx, Shape{param_sizes[10]/NUM_PARAMETER_LAYERS}, kf32));
+      params->fcb.push_back(createTensor(ctx, Shape{param_sizes[11]/NUM_PARAMETER_LAYERS}, kf32));
+      params->fcprojw.push_back(createTensor(ctx, Shape{param_sizes[12]/NUM_PARAMETER_LAYERS}, kf32));
+      params->fcprojb.push_back(createTensor(ctx, Shape{param_sizes[13]/NUM_PARAMETER_LAYERS}, kf32));
+    }
+    params->lnfw = createTensor(ctx, Shape{param_sizes[14]}, kf32);
+    params->lnfb = createTensor(ctx, Shape{param_sizes[15]}, kf32);
+}
+
+
+#define NUM_ACTIVATION_TENSORS 23
+typedef struct {
+    Tensor encoded; // (B, T, C)
+    std::vector<Tensor> ln1; // (L, B, T, C)
```

Comment:
Could we use std::array instead with NUM_ACTIVATION_TENSORS statically allocated?

Reply:
It may be confusing, but NUM_ACTIVATION_TENSORS is the number of variables like `encoded` and `ln1`, not the size of the vector.